Add `global-search-filters` feature

kidonng opened the pull request:
<!--

Thanks for contributing! 🍄 Do not ignore this template, plz.

1. Does this PR close/fix an existing issue? Write something like `Closes #10`

Help us test and visualize this PR:

2. What pages does this PR affect? Include some REAL URLs where you tested the code
3. If applicable, add demonstrative screenshots or gifs

Lastly:

4. Open the PR as draft and review it yourself first. Fix what you find and explain weird code, if necessary.

-->

Resolve #3613

Better names welcome


## Test URLs

- Sidebar with only categories: https://github.com/search?q=author%3A%40me&type=commits
- Sidebar with categories and languages: https://github.com/search?q=waifu2x-converter-cpp
- Sidebar with  categories, states and languages: https://github.com/search?q=author%3A%40me&type=issues


## Screenshot

![image](https://user-images.githubusercontent.com/44045911/112261285-88a66c80-8ca6-11eb-82cb-933b72c57abd.png)

kidonng commented:
Actually I have some questions unsolved:

- Should we show a link with `fork: true` if the user doesn't use the dropdown, for example if they clicked a link?
- Should we remove `fork: true` if the user switch to other sorting options?

kidonng commented:
I think I will go for the alternate option mentioned [here](https://github.com/sindresorhus/refined-github/issues/3613#issuecomment-796695408): add a checkbox to include forks

Take waifu2x-converter-cpp fork:true as an example, you already get good enough results when including forks. You don't have to change sorting criteria.

fregante commented on source/features/show-forks-in-search.css:
```diff
@@ -0,0 +1,3 @@
+.rgh-show-forks-in-search-checkbox {
+	vertical-align: middle;
```
I think that there's a GitHub class already for this, like `v-align-middle`

fregante commented on source/features/show-forks-in-search.tsx:
```diff
@@ -0,0 +1,33 @@
+
+function init(): void {
+	const includeForks = new SearchQuery(location.search).includes('fork:true');
+	const checkbox = <label className="float-right"><input className="rgh-show-forks-in-search-checkbox mr-1" type="checkbox" checked={includeForks}/>Include forks</label>;
```
```suggestion
	const checkbox = (
		<label className="float-right">
			  <input className="rgh-show-forks-in-search-checkbox mr-1" type="checkbox" checked={includeForks}/>
			  Include forks
		</label>
	);
```

kidonng replied:
I remember linter would yell at me for this before, but it didn't so I passed it for now 😄

fregante replied:
If you find incomprehensible/difficult build/lint errors you can ask for help, as long as the logic works.

fregante commented on source/features/show-forks-in-search.tsx:
```diff
@@ -0,0 +1,33 @@
+import SearchQuery from '../github-helpers/search-query';
+
+function init(): void {
+	const includeForks = new SearchQuery(location.search).includes('fork:true');
```
:tada:

fregante commented on source/features/show-forks-in-search.tsx:
```diff
@@ -0,0 +1,33 @@
+function init(): void {
+	const includeForks = new SearchQuery(location.search).includes('fork:true');
+	const checkbox = <label className="float-right"><input className="rgh-show-forks-in-search-checkbox mr-1" type="checkbox" checked={includeForks}/>Include forks</label>;
+	checkbox.addEventListener('click', () => {
```
If the page has PJAX enabled, this listener will be lost on `history.back()`. Try and if it breaks, it needs to use `delegate`

Also it should probably be `change` instead of `click`

kidonng replied:
> Also it should probably be `change` instead of `click`

I noticed this, but haven't fixed it yet. Well it says `checkbox` but it is a in fact a label 😅

fregante commented on source/features/show-forks-in-search.tsx:
```diff
@@ -0,0 +1,33 @@
+	const checkbox = <label className="float-right"><input className="rgh-show-forks-in-search-checkbox mr-1" type="checkbox" checked={includeForks}/>Include forks</label>;
+	checkbox.addEventListener('click', () => {
+		for (const link of select.all<HTMLAnchorElement>('.codesearch-results .select-menu-item')) {
+			if (includeForks) {
+				new SearchQuery(link).remove('fork:true');
+			} else {
+				new SearchQuery(link).add('fork:true');
+			}
+
+			if (link.getAttribute('aria-checked') === 'true') {
+				link.click();
+			}
```
Why change all links in the loop if we're about to lose the dom with `link.click()`?

kidonng replied:
I was being dumb here, possibly I can just get rid of the loop.

fregante reviewed (changes requested):
If I understand the code correctly, even if you correctly detect `forks:true` on load and mark the checkbox as `checked`, the links in the dropdown aren't actually updated in this case.

Example:

1. Visit the search page
2. Click the checkbox
3. The page reloads, right?
4. Now select another item in the list
5. The page reloads, but `forks:true` is lost

---

There's also a second issue here, I think: It's being shown as a checkbox but clicking it will navigate to the new search page immediately. Checkboxes shouldn't reload pages.

Also, I think this is a bad place for the checkbox since it's a sort menu, not a filter menu.

kidonng commented:
> There's also a second issue here, I think: It's being shown as a checkbox but clicking it will navigate to the new search page immediately. Checkboxes shouldn't reload pages.

Well, I just think it's not obvious you have to click the link again to see the new results.

> Also, I think this is a bad place for the checkbox since it's a sort menu, not a filter menu.

Tried to place it next to the dropdown, but the container seems to need a lot of modifications.

fregante commented:
> Tried to place it next to the dropdown, but I'm bad at positioning it (and make it responsive). Suggestions welcome.

Filters are in the sidebar. We can add a whole block like this one, even if it's just one "Forks" link inside it (with an X to remove the filter, just like the Languages box)

<img width="258" alt="Screen Shot" src="https://user-images.githubusercontent.com/1402241/111139567-e6b8ae80-8546-11eb-8ac8-70e54167f499.png">

Thoughts?

kidonng commented:
> Filters are in the sidebar.

Wow I totally forget that.

> We can add a whole block like this one, even if it's just one "Forks" link inside it (with an X to remove the filter, just like the Languages box)

Hmm actually this opens new possibilities. Maybe add some other filters as well? At least I found `is:private` useful too.

fregante commented:
> `user: myself` useful too

You can already start the search with your own user

<img width="559" alt="Screen Shot" src="https://user-images.githubusercontent.com/1402241/111145788-45cdf180-854e-11eb-982a-e6b5a81acdb8.png">

> Maybe add some other filters as well

I'm -0 on this.


> Possible design

That's it, as long as it doesn't actually replace the Languages one

kidonng commented:
> I'm -0 on this.

I won't turn it into another long list of filters, and after reading the list of available filters I end up with only those I mentioned.

> That's it, as long as it doesn't actually replace the Languages one

Why would I do that

fregante commented:
> > That's it, as long as it doesn't actually replace the Languages one
> 
> Why would I do that

That's what the screenshot looks like 😃

fregante commented on source/features/search-filters.tsx:
```diff
@@ -0,0 +1,48 @@
+		return <li>{item}</li>;
+	});
+
+	select('.menu')!.nextElementSibling!.after(
```
Selector wait too broad

fregante commented on source/features/search-filters.tsx:
```diff
@@ -0,0 +1,48 @@
+		['Yours', `user:${getUsername()}`]
+	].map(([name, filter]) => {
+		const item = <a className="filter-item" href={location.href}>{name}</a>;
+		// @ts-expect-error
```
This should be fixed before merging. Let it fail or else we don't even see the error and help
```suggestion
```

kidonng replied:
`item` is `JSX.Element` while `SearchQuery` expects `HTMLAnchorElement`.

fregante commented on source/features/search-filters.tsx:
```diff
@@ -0,0 +1,48 @@
+	const items = [
+		['Forks', 'fork:true'],
+		['Private', 'is:private'],
+		['Yours', `user:${getUsername()}`]
```
```suggestion
		['Yours', `user:@me`]
```

kidonng replied:
For some reason GitHub doesn't show code search results when using `user:@me`:

<details>
<summary><code>user:@me</code></summary>

![image](https://user-images.githubusercontent.com/44045911/111408437-43b39080-8710-11eb-932c-b9e63e9e8a85.png)

</details>

<details>
<summary><code>user:username</code></summary>

![image](https://user-images.githubusercontent.com/44045911/111408484-529a4300-8710-11eb-9c6c-73263d6abd20.png)

</details>

fregante commented on source/features/search-filters.tsx:
```diff
@@ -0,0 +1,48 @@
+		['Forks', 'fork:true'],
+		['Private', 'is:private'],
+		['Yours', `user:${getUsername()}`]
+	].map(([name, filter]) => {
```
```diff
- .map()
+ for of
```

and then
```diff
- return
+ items.push()
```

fregante commented on readme.md:
```diff
@@ -315,6 +315,7 @@ Thanks for contributing! 🦋🙌
 - [](# "improve-shortcut-help") [Shows all of Refined GitHub’s new keyboard shortcuts in the help modal (<kbd>?</kbd> hotkey).](https://user-images.githubusercontent.com/29176678/36999174-9f07d33e-20bf-11e8-83e3-b3a9908a4b5f.png)
 - [](# "stop-redirecting-in-notification-bar") [Stops redirecting to notification inbox from notification bar actions while holding <kbd>Alt</kbd>.](https://user-images.githubusercontent.com/202916/80318782-c38cef80-880c-11ea-9226-72c585f42a51.png)
 - [](# "hide-zero-packages") [Hides the `Packages` tab in user profiles if it’s empty.](https://user-images.githubusercontent.com/46634000/101151879-9ade9b00-3622-11eb-9755-50ed54099e02.png)
+- [](# "global-search-filters") [Adds useful search filters in search page.](https://user-images.githubusercontent.com/44045911/111149280-a6e5db80-85c7-11eb-9e61-4badb9d3c685.png)
```
```suggestion
- [](# "global-search-filters") [Adds search filters in global search page: forks, private repos, own repos.](https://user-images.githubusercontent.com/44045911/111149280-a6e5db80-85c7-11eb-9e61-4badb9d3c685.png)
```

fregante commented on source/features/global-search-filters.tsx:
```diff
@@ -0,0 +1,49 @@
+		items.push(<li>{item}</li>);
+	}
+
+	select('#js-pjax-container .menu')!.nextElementSibling!.after(
```
This adds it in place of the languages; it should be added below it to avoid moving the UI too much. Also it's best to use a better selector instead of `nextElementSibling`

fregante reviewed (approved):
Untested but it looks correct other than this

yakov116 commented on source/features/global-search-filters.tsx:
```diff
@@ -0,0 +1,49 @@
+	const filters = [
+		['Forks', 'fork:true'],
+		['Private', 'is:private'],
+		['Yours', `user:${getUsername()}`]
```
```suggestion
		['Yours', 'user:' + getUsername()]
```

fregante replied:
Don't let Sindre see this 😃

yakov116 commented on source/features/global-search-filters.tsx:
```diff
@@ -0,0 +1,49 @@
+		items.push(<li>{item}</li>);
+	}
+
+	select('#js-pjax-container .menu ~ .border')!.after(
```
Could we make this not have to wait for domReady by using element ready?

yakov116 commented:
@kidonng no pressure. But there are just 2 more things to get this merged

kidonng commented:
Sorry, will visit these today or tomorrow.

kidonng commented:
Resolved reviews and added more test URLs. I also added a new **Authored** filter `author:@me`, which is useful for e.g. search commits or conversations authored by yourself.

yakov116 commented:
Thanks @kidonng

641210 reviewed (commented):
****
